File: src/main/java/com/goodhuddle/huddle/service/PageService.java
/*
 * (C) Copyright ${year} Nuxeo SA (http://nuxeo.com/) and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 2.1 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl-2.1.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 */

package com.goodhuddle.huddle.service;

import com.goodhuddle.huddle.domain.Page;
import com.goodhuddle.huddle.service.exception.PageSlugExistsException;
import com.goodhuddle.huddle.service.request.page.CreatePageRequest;
import com.goodhuddle.huddle.service.request.page.UpdatePageRequest;

import java.io.File;
import java.io.IOException;
import java.util.List;

public interface PageService {

    List<Page> getPages();

    Page getPage(long pageId);

    Page getPageForSlug(String slug);

    Page createPage(CreatePageRequest request) throws PageSlugExistsException;

    Page updatePage(UpdatePageRequest request) throws PageSlugExistsException;

    void deletePage(long pageId);

    File saveAttachment(long pageId, String fileName, byte[] data) throws IOException;

    File getAttachment(long pageId, String fileName) throws IOException;

}

File: src/main/resources/static/admin/service/page-service.js
/*
 * (C) Copyright ${year} Nuxeo SA (http://nuxeo.com/) and others.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the GNU Lesser General Public License
 * (LGPL) version 2.1 which accompanies this distribution, and is available at
 * http://www.gnu.org/licenses/lgpl-2.1.html
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 */

'use strict';

angular.module('huddleAdmin.service.page', ['ngResource'])

    .service('PageService', function ($resource, $http) {

        return $resource('/api/page/:id', {id: '@id'},
            {
                get: {
                    method: 'GET',
                    transformResponse: $http.defaults.transformResponse.concat([
                        function (page, headersGetter) {
                            angular.forEach(page.content.rows, function(row, index) {
                                var start = 0;
                                for (var i = 0; i < row.cells.length; i++) {
                                    row.cells[i].start = start;
                                    start += row.cells[i].width;
                                }
                            });
                            return page;
                        }
                    ])
                },

                update: {
                    method: 'PUT'
                },

                preview: {
                    method: 'POST',
                    url: '/page/preview'
                },

                list: {
                    method: 'POST',
                    url: '/api/search'
                }
            });
    });
